Null pointer exception when starting ad-hoc snapshot with wrong content [DBZ-6352]

Migrated from [DBZ-6352](https://issues.redhat.com/browse/DBZ-6352)

h2. What Debezium connector do you use and what version?

Debezium PostgreSQL 2.0.0.Final, but the code has not changed

h2. What is the connector configuration?


{code:json}
{
    "connector.class": "io.debezium.connector.postgresql.PostgresConnector",
    "incremental.snapshot.chunk.size": "5120",
    "slot.name": "debezium_cdc",
    "tasks.max": "1",
    "publication.name": "debezium_cdc",
    "value.converter.enhanced.avro.schema.support": "true",
    "value.converter.basic.auth.credentials.source": "USER_INFO",
    "database.sslmode": "require",
    "topic.prefix": "registry",
    "signal.data.collection": "public.debezium_signals",
    "value.converter": "io.confluent.connect.avro.AvroConverter",
    "errors.log.enable": "true",
    "key.converter": "io.confluent.connect.avro.AvroConverter",
    "publication.autocreate.mode": "disabled",
    "database.dbname": "snykregistry",
    "database.user": "username",
    "database.server.name": "registry",
    "key.converter.enhanced.avro.schema.support": "true",
    "database.port": "5432",
    "plugin.name": "pgoutput",
    "key.converter.basic.auth.user.info": "userinfo",
    "value.converter.schema.registry.url": "schema-registry-url",
    "column.exclude.list": "public.a_table.a_column",
    "value.converter.basic.auth.user.info": "user-info",
    "database.hostname": "db-hostname",
    "database.password": "password",
    "name": "debezium-cdc",
    "errors.tolerance": "all",
    "skipped.operations": "t",
    "table.include.list": "a list of tables",
    "key.converter.schema.registry.url": "url",
    "key.converter.basic.auth.credentials.source": "USER_INFO",
    "snapshot.mode": "never"
}
{code}

h2. What is the captured database version and mode of depoyment?

GCP PostgreSQL 14 (really doesn't matter)

h2. What behaviour do you expect?

An invalid `data` column in the signals table will trigger an error in the logs, but WAL log processing will continue.

h2. What behaviour do you see?

When inserting an invalid `data` column in the signals table, Debezium will fail and completely stop processing messages. Restarting the connector won't help.

The data column had a nested array (this was an error on our side). For example:
```{"data-collections":[["tablea","tableb"]]}```

h2. Do you see the same behaviour using the latest relesead Debezium version?

Can't try easily, but code is the same https://github.com/debezium/debezium/blob/db4c2fd91844f3401bc34d62a8033892d5b52987/debezium-core/src/main/java/io/debezium/pipeline/signal/ExecuteSnapshot.java#L71-L73

h2. Do you have the connector logs, ideally from start till finish?

{noformat}
org.apache.kafka.connect.errors.ConnectException: An exception occurred in the change event producer. This connector will be stopped.
	at io.debezium.pipeline.ErrorHandler.setProducerThrowable(ErrorHandler.java:53)
	at io.debezium.connector.postgresql.PostgresStreamingChangeEventSource.execute(PostgresStreamingChangeEventSource.java:180)
	at io.debezium.connector.postgresql.PostgresStreamingChangeEventSource.execute(PostgresStreamingChangeEventSource.java:41)
	at io.debezium.pipeline.ChangeEventSourceCoordinator.streamEvents(ChangeEventSourceCoordinator.java:174)
	at io.debezium.pipeline.ChangeEventSourceCoordinator.executeChangeEventSources(ChangeEventSourceCoordinator.java:141)
	at io.debezium.pipeline.ChangeEventSourceCoordinator.lambda$start$0(ChangeEventSourceCoordinator.java:109)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.apache.kafka.connect.errors.ConnectException: Error while processing event at offset {transaction_id=null, lsn_proc=40020183155160, lsn_commit=40020183155160, lsn=40020183155160, txId=714621851, ts_usec=1681843474910582}
	at io.debezium.pipeline.EventDispatcher.dispatchDataChangeEvent(EventDispatcher.java:261)
	at io.debezium.connector.postgresql.PostgresStreamingChangeEventSource.lambda$processMessages$0(PostgresStreamingChangeEventSource.java:275)
	at io.debezium.connector.postgresql.connection.pgoutput.PgOutputMessageDecoder.decodeInsert(PgOutputMessageDecoder.java:413)
	at io.debezium.connector.postgresql.connection.pgoutput.PgOutputMessageDecoder.processNotEmptyMessage(PgOutputMessageDecoder.java:191)
	at io.debezium.connector.postgresql.connection.AbstractMessageDecoder.processMessage(AbstractMessageDecoder.java:41)
	at io.debezium.connector.postgresql.connection.PostgresReplicationConnection$1.deserializeMessages(PostgresReplicationConnection.java:504)
	at io.debezium.connector.postgresql.connection.PostgresReplicationConnection$1.readPending(PostgresReplicationConnection.java:496)
	at io.debezium.connector.postgresql.PostgresStreamingChangeEventSource.processMessages(PostgresStreamingChangeEventSource.java:215)
	at io.debezium.connector.postgresql.PostgresStreamingChangeEventSource.execute(PostgresStreamingChangeEventSource.java:177)
	... 9 more
Caused by: java.lang.NullPointerException
	at io.debezium.pipeline.signal.ExecuteSnapshot.lambda$getDataCollections$0(ExecuteSnapshot.java:70)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.debezium.pipeline.signal.ExecuteSnapshot.getDataCollections(ExecuteSnapshot.java:71)
	at io.debezium.pipeline.signal.ExecuteSnapshot.arrived(ExecuteSnapshot.java:44)
	at io.debezium.pipeline.signal.Signal.process(Signal.java:138)
	at io.debezium.pipeline.signal.Signal.process(Signal.java:176)
	at io.debezium.pipeline.EventDispatcher$2.changeRecord(EventDispatcher.java:235)
	at io.debezium.relational.RelationalChangeRecordEmitter.emitCreateRecord(RelationalChangeRecordEmitter.java:79)
	at io.debezium.relational.RelationalChangeRecordEmitter.emitChangeRecords(RelationalChangeRecordEmitter.java:47)
	at io.debezium.connector.postgresql.PostgresChangeRecordEmitter.emitChangeRecords(PostgresChangeRecordEmitter.java:92)
	at io.debezium.pipeline.EventDispatcher.dispatchDataChangeEvent(EventDispatcher.java:224)
{noformat}

h2. How to reproduce the issue using our [tutorial|https://github.com/debezium/debezium-examples/tree/main/tutorial] deployment?

Create connector that supports signals, insert message like 

{noformat}
id='09ac1a68-df2e-41e3-8757-28c266b2632a', type='execute-snapshot', data='{"data-collections":[["tablea"]]}'
{noformat}

Any help to get the connector running again would be appreciated

--- a/debezium-connector-common/src/main/java/io/debezium/pipeline/signal/SignalProcessor.java
+++ b/debezium-connector-common/src/main/java/io/debezium/pipeline/signal/SignalProcessor.java
@@ -267,6 +267,10 @@ public class SignalProcessor<P extends Partition, O extends OffsetContext> {
 
         // Extract database name from first data collection
         for (Array.Entry entry : dataCollectionsArray) {
+            if (!entry.getValue().isString()) {
+                LOGGER.debug("Skipping non-string data-collections element while resolving the signal partition");
+                continue;
+            }
             String dataCollection = entry.getValue().asString().trim();
 
             String[] parts = dataCollection.split(POINT_REGEX);

--- a/debezium-connector-common/src/main/java/io/debezium/pipeline/signal/actions/AbstractSnapshotSignal.java
+++ b/debezium-connector-common/src/main/java/io/debezium/pipeline/signal/actions/AbstractSnapshotSignal.java
@@ -5,6 +5,7 @@
  */
 package io.debezium.pipeline.signal.actions;
 
+import java.util.List;
 import java.util.Optional;
 
 import org.slf4j.Logger;
@@ -31,6 +32,16 @@ public abstract class AbstractSnapshotSignal<P extends Partition> implements Sig
     public enum SnapshotType {
         INCREMENTAL,
         BLOCKING
+    }
+
+    /**
+     * Result of parsing the {@code data-collections} field of a snapshot signal.
+     *
+     * @param valid whether the signal can be acted upon; {@code false} when the field is present but malformed
+     *            (for example a non-string element), so the signal must be skipped
+     * @param collections the parsed collections, or {@code null} when the field is missing or empty
+     */
+    public record DataCollections(boolean valid, List<String> collections) {
     }
 
     public static SnapshotType getSnapshotType(Document data) {

--- a/debezium-connector-common/src/main/java/io/debezium/pipeline/signal/actions/snapshotting/ExecuteSnapshot.java
+++ b/debezium-connector-common/src/main/java/io/debezium/pipeline/signal/actions/snapshotting/ExecuteSnapshot.java
@@ -52,10 +52,11 @@ public class ExecuteSnapshot<P extends Partition> extends AbstractSnapshotSignal
     @Override
     public boolean arrived(SignalPayload<P> signalPayload) throws InterruptedException {
 
-        final List<String> dataCollections = getDataCollections(signalPayload.data);
-        if (dataCollections == null) {
+        final DataCollections dataCollectionsResult = getDataCollections(signalPayload.data);
+        if (!dataCollectionsResult.valid()) {
             return false;
         }
+        final List<String> dataCollections = dataCollectionsResult.collections();
         SnapshotType type = getSnapshotType(signalPayload.data);
 
         List<AdditionalCondition> additionalConditions = getAdditionalConditions(signalPayload.data, type);
@@ -100,18 +101,24 @@ public class ExecuteSnapshot<P extends Partition> extends AbstractSnapshotSignal
                 .build();
     }
 
-    public static List<String> getDataCollections(Document data) {
+    public static DataCollections getDataCollections(Document data) {
 
         final Array dataCollectionsArray = data.getArray(FIELD_DATA_COLLECTIONS);
         if (dataCollectionsArray == null || dataCollectionsArray.isEmpty()) {
             LOGGER.warn(
                     "Execute snapshot signal '{}' has arrived but the requested field '{}' is missing from data or is empty",
                     data, FIELD_DATA_COLLECTIONS);
-            return null;
+            return new DataCollections(false, null);
         }
-        return dataCollectionsArray.streamValues()
+        if (dataCollectionsArray.streamValues().anyMatch(v -> !v.isString())) {
+            LOGGER.warn(
+                    "Execute snapshot signal '{}' has arrived but the requested field '{}' contains a non-string data collection",
+                    data, FIELD_DATA_COLLECTIONS);
+            return new DataCollections(false, null);
+        }
+        return new DataCollections(true, dataCollectionsArray.streamValues()
                 .map(v -> v.asString().trim())
-                .collect(Collectors.toList());
+                .collect(Collectors.toList()));
     }
 
     public static Optional<String> getSurrogateKey(Document data) {

--- a/debezium-connector-common/src/main/java/io/debezium/pipeline/signal/actions/snapshotting/StopSnapshot.java
+++ b/debezium-connector-common/src/main/java/io/debezium/pipeline/signal/actions/snapshotting/StopSnapshot.java
@@ -42,7 +42,11 @@ public class StopSnapshot<P extends Partition> extends AbstractSnapshotSignal<P>
 
     @Override
     public boolean arrived(SignalPayload<P> signalPayload) throws InterruptedException {
-        final List<String> dataCollections = getDataCollections(signalPayload.data);
+        final DataCollections dataCollectionsResult = getDataCollections(signalPayload.data);
+        if (!dataCollectionsResult.valid()) {
+            return false;
+        }
+        final List<String> dataCollections = dataCollectionsResult.collections();
         final SnapshotType type = getSnapshotType(signalPayload.data);
 
         LOGGER.info("Requested stop of snapshot '{}' for data collections '{}'", type, dataCollections);
@@ -56,13 +60,20 @@ public class StopSnapshot<P extends Partition> extends AbstractSnapshotSignal<P>
         return true;
     }
 
-    public static List<String> getDataCollections(Document data) {
+    public static DataCollections getDataCollections(Document data) {
         final Array dataCollectionsArray = data.getArray(FIELD_DATA_COLLECTIONS);
         if (dataCollectionsArray == null || dataCollectionsArray.isEmpty()) {
-            return null;
+            // A missing or empty data-collections means the entire in-progress snapshot is stopped.
+            return new DataCollections(true, null);
         }
-        return dataCollectionsArray.streamValues()
+        if (dataCollectionsArray.streamValues().anyMatch(v -> !v.isString())) {
+            LOGGER.warn(
+                    "Stop snapshot signal '{}' has arrived but the requested field '{}' contains a non-string data collection",
+                    data, FIELD_DATA_COLLECTIONS);
+            return new DataCollections(false, null);
+        }
+        return new DataCollections(true, dataCollectionsArray.streamValues()
                 .map(v -> v.asString().trim())
-                .collect(Collectors.toList());
+                .collect(Collectors.toList()));
     }
 }

--- a/debezium-connector-common/src/test/java/io/debezium/pipeline/signal/actions/snapshotting/SnapshotSignalArrivedTest.java
+++ b/debezium-connector-common/src/test/java/io/debezium/pipeline/signal/actions/snapshotting/SnapshotSignalArrivedTest.java
@@ -1,0 +1,89 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.pipeline.signal.actions.snapshotting;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.ArgumentMatchers.isNull;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import io.debezium.document.Document;
+import io.debezium.document.DocumentReader;
+import io.debezium.pipeline.EventDispatcher;
+import io.debezium.pipeline.signal.SignalPayload;
+import io.debezium.pipeline.source.snapshot.incremental.IncrementalSnapshotChangeEventSource;
+import io.debezium.pipeline.spi.OffsetContext;
+import io.debezium.pipeline.spi.Partition;
+
+/**
+ * Verifies the behaviour of {@code arrived()} for the snapshot signals, in particular that a malformed
+ * {@code data-collections} is skipped rather than acted upon. A malformed stop signal must not escalate to
+ * stopping the entire in-progress snapshot, while a missing/empty one still does.
+ *
+ * @see <a href="https://github.com/debezium/dbz/issues/798">DBZ-6352</a>
+ */
+class SnapshotSignalArrivedTest {
+
+    private final DocumentReader reader = DocumentReader.defaultReader();
+
+    @SuppressWarnings("unchecked")
+    private final EventDispatcher<Partition, ?> dispatcher = mock(EventDispatcher.class);
+    @SuppressWarnings("unchecked")
+    private final IncrementalSnapshotChangeEventSource<Partition, ?> incrementalSource = mock(IncrementalSnapshotChangeEventSource.class);
+
+    @BeforeEach
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    void setUp() {
+        when(dispatcher.getIncrementalSnapshotChangeEventSource()).thenReturn((IncrementalSnapshotChangeEventSource) incrementalSource);
+    }
+
+    private SignalPayload<Partition> payload(String json) throws IOException {
+        final Document data = reader.read(json);
+        return new SignalPayload<>(mock(Partition.class), "signal-id", "stop-snapshot", data, mock(OffsetContext.class), Map.of());
+    }
+
+    @Test
+    void stopSnapshotShouldNotStopAnythingForMalformedDataCollections() throws IOException, InterruptedException {
+        final StopSnapshot<Partition> action = new StopSnapshot<>(dispatcher);
+
+        final boolean result = action.arrived(payload("{\"data-collections\": [[\"schema.table1\"]]}"));
+
+        assertThat(result).isFalse();
+        // The whole point of the fix: a malformed stop signal must NOT escalate to stopping the snapshot.
+        verify(incrementalSource, never()).requestStopSnapshot(any(), any(), any(), any());
+    }
+
+    @Test
+    void stopSnapshotShouldStopAllForMissingDataCollections() throws IOException, InterruptedException {
+        final StopSnapshot<Partition> action = new StopSnapshot<>(dispatcher);
+
+        final boolean result = action.arrived(payload("{}"));
+
+        assertThat(result).isTrue();
+        // Missing data-collections is a valid request to stop the entire snapshot: null flows through to stop-all.
+        verify(incrementalSource).requestStopSnapshot(any(), any(), any(), isNull());
+    }
+
+    @Test
+    void stopSnapshotShouldStopSpecificCollections() throws IOException, InterruptedException {
+        final StopSnapshot<Partition> action = new StopSnapshot<>(dispatcher);
+
+        final boolean result = action.arrived(payload("{\"data-collections\": [\"schema.table1\"]}"));
+
+        assertThat(result).isTrue();
+        verify(incrementalSource).requestStopSnapshot(any(), any(), any(), eq(java.util.List.of("schema.table1")));
+    }
+}

--- a/debezium-connector-common/src/test/java/io/debezium/pipeline/signal/actions/snapshotting/SnapshotSignalDataCollectionsTest.java
+++ b/debezium-connector-common/src/test/java/io/debezium/pipeline/signal/actions/snapshotting/SnapshotSignalDataCollectionsTest.java
@@ -1,0 +1,94 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.pipeline.signal.actions.snapshotting;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.IOException;
+
+import org.junit.jupiter.api.Test;
+
+import io.debezium.document.Document;
+import io.debezium.document.DocumentReader;
+import io.debezium.junit.logging.LogInterceptor;
+import io.debezium.pipeline.signal.actions.AbstractSnapshotSignal.DataCollections;
+
+/**
+ * Verifies that the {@code data-collections} parsing shared by the snapshot signals rejects malformed
+ * (non-string) entries instead of throwing a {@link NullPointerException} that stops the connector.
+ *
+ * @see <a href="https://github.com/debezium/dbz/issues/798">DBZ-6352</a>
+ */
+class SnapshotSignalDataCollectionsTest {
+
+    private final DocumentReader reader = DocumentReader.defaultReader();
+
+    private Document parse(String json) throws IOException {
+        return reader.read(json);
+    }
+
+    @Test
+    void executeSnapshotShouldParseStringDataCollections() throws IOException {
+        DataCollections result = ExecuteSnapshot.getDataCollections(parse("{\"data-collections\": [\"schema.table1\", \"schema.table2\"]}"));
+
+        assertThat(result.valid()).isTrue();
+        assertThat(result.collections()).containsExactly("schema.table1", "schema.table2");
+    }
+
+    @Test
+    void executeSnapshotShouldBeInvalidForNestedArrayDataCollections() throws IOException {
+        LogInterceptor logInterceptor = new LogInterceptor(ExecuteSnapshot.class);
+
+        DataCollections result = ExecuteSnapshot.getDataCollections(parse("{\"data-collections\": [[\"schema.table1\", \"schema.table2\"]]}"));
+
+        assertThat(result.valid()).isFalse();
+        assertThat(logInterceptor.containsWarnMessage("contains a non-string data collection")).isTrue();
+    }
+
+    @Test
+    void executeSnapshotShouldBeInvalidWhenAnyDataCollectionIsNotAString() throws IOException {
+        DataCollections result = ExecuteSnapshot.getDataCollections(parse("{\"data-collections\": [\"schema.table1\", 42]}"));
+
+        assertThat(result.valid()).isFalse();
+    }
+
+    @Test
+    void executeSnapshotShouldBeInvalidWhenDataCollectionsMissingOrEmpty() throws IOException {
+        assertThat(ExecuteSnapshot.getDataCollections(parse("{}")).valid()).isFalse();
+        assertThat(ExecuteSnapshot.getDataCollections(parse("{\"data-collections\": []}")).valid()).isFalse();
+    }
+
+    @Test
+    void stopSnapshotShouldParseStringDataCollections() throws IOException {
+        DataCollections result = StopSnapshot.getDataCollections(parse("{\"data-collections\": [\"schema.table1\"]}"));
+
+        assertThat(result.valid()).isTrue();
+        assertThat(result.collections()).containsExactly("schema.table1");
+    }
+
+    @Test
+    void stopSnapshotShouldBeInvalidForNestedArrayDataCollections() throws IOException {
+        LogInterceptor logInterceptor = new LogInterceptor(StopSnapshot.class);
+
+        DataCollections result = StopSnapshot.getDataCollections(parse("{\"data-collections\": [[\"schema.table1\"]]}"));
+
+        assertThat(result.valid()).isFalse();
+        assertThat(logInterceptor.containsWarnMessage("contains a non-string data collection")).isTrue();
+    }
+
+    @Test
+    void stopSnapshotShouldStopAllWhenDataCollectionsMissingOrEmpty() throws IOException {
+        // A missing or empty data-collections is a valid request to stop the entire snapshot, distinct from a
+        // malformed one; both must stay separable so a malformed signal is skipped rather than escalating to stop-all.
+        DataCollections missing = StopSnapshot.getDataCollections(parse("{}"));
+        assertThat(missing.valid()).isTrue();
+        assertThat(missing.collections()).isNull();
+
+        DataCollections empty = StopSnapshot.getDataCollections(parse("{\"data-collections\": []}"));
+        assertThat(empty.valid()).isTrue();
+        assertThat(empty.collections()).isNull();
+    }
+}

